### PR TITLE
Add candidate tracker and progress bar to autotune.py

### PR DESCRIPTION
**Pull request by RattataKing**
autotune.py: 
1. Added dataclass `CandidateTracker`
2. Added progress bar for all candidates processing functions, excluding `generate_candidates()`

tune.py:
1. Fixed incompatibilities with `autotune.sh` (directory existed issue)

benchmark_dispatch.sh:
1. Edited a dir name (`compiled/failed` -> `compiled/benchmark_failed`)

### CODE REVIEW

**Review by kuhar — COMMENTED** [review]
Looks good, just please format the code and make the PR title and description more descriptive. Some pointers here: https://www.gitkraken.com/learn/git/best-practices/git-commit-message.

File: tuning/autotune.py [hunk 1]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+        # The rest of the filename after the numeric part
+        remaining_part = path.stem[len(match.group(0)):]
+    else:
+        numeric_part = float('inf')
+        remaining_part = path.stem
+    return (numeric_part, remaining_part)
+
+
+def calculate_md5(file_path):
+    """calculate the MD5 hash"""
```
Comment (kuhar) [inline tuning/autotune.py:253]: This comment doesn't add anything on top of the function name
```suggestion
```

File: tuning/autotune.py [hunk 2]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+        numeric_part = int(match.group(1))
+        # The rest of the filename after the numeric part
+        remaining_part = path.stem[len(match.group(0)):]
+    else:
+        numeric_part = float('inf')
+        remaining_part = path.stem
+    return (numeric_part, remaining_part)
+
+
+def calculate_md5(file_path):
```
Comment (kuhar) [inline tuning/autotune.py:252]: Add type hints

File: tuning/autotune.py [hunk 3]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+        # Use tqdm to create a progress bar
+        with tqdm(total=len(task_list)) as pbar:
+            # Use imap_unordered to asynchronously execute the worker function on each task
+            for result in worker_pool.imap_unordered(function, task_list):
+                pbar.update(1)  # Update progress bar
+                results.append(result)
+    return results
+
+
+def numerical_sort_key(path):
```
Comment (kuhar) [inline tuning/autotune.py:237]: Add type hints

File: tuning/autotune.py [hunk 4]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+    """Define a sort key function that splits the filename into a numeric and a string part"""
+    """Order: 0 | 0_a | 0_b | 1 | 1_a | 2"""
```
Comment (kuhar) [inline tuning/autotune.py:239]: nit: merge these two comments?

File: tuning/autotune.py [hunk 5]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+) -> subprocess.CompletedProcess:
+    """pool.imap_unordered can't iterate an iterable of iterables input, this function helps dividing arguments"""
+    args, command, check = task_tuple
+    return run_command(args, command, check)
+
+
+def multiprocess_progress_wrapper(
+    num_worker: int,
+    task_list: list,
+    function,
```
Comment (kuhar) [inline tuning/autotune.py:218]: type?

File: tuning/autotune.py [hunk 6]
```diff
@@ -25,10 +29,26 @@
...
+    configuration: tune.Configuration = None
+    compilation_successful: bool = None
+    compiled_vmfb_path: str = None
+    first_benchmark_time: float = None
+    unet_candidate_path: str = None
+    unet_vmfb_hash: str = None
+    unet_benchmark_time: float = None
+
+
+Candidate_Tracker: list[TuningCandidate] = []
```
Comment (kuhar) [inline tuning/autotune.py:50]: Can we avoid using global variables? These make it difficult to understand where the program state changes and are hard to test.

File: tuning/autotune.py [hunk 7]
```diff
@@ -183,6 +204,60 @@ def run_command(
...
+        remaining_part = path.stem
+    return (numeric_part, remaining_part)
+
+
+def calculate_md5(file_path):
+    """calculate the MD5 hash"""
+    md5 = hashlib.md5()
+    with open(file_path, 'rb') as f:
+        for chunk in iter(lambda: f.read(4096), b''):
+            md5.update(chunk)
```
Comment (kuhar) [inline tuning/autotune.py:257]: Do we need to split this in chunks? This example with sha256 doesn't do that: https://docs.python.org/3/library/hashlib.html#file-hashing

(Either sha256 or md5 are fine)

Comment (RattataKing) [reply tuning/autotune.py:256]: I searched online `hashlib.file_digest` is a higher-level function introduced in Python 3.11, it also reads the file in chunks:
https://github.com/python/cpython/blob/030b452e34bbb0096acacb70a31915b9590c8186/Lib/hashlib.py#L230

this function looks premium, i will use it

Comment (RattataKing) [reply tuning/autotune.py:256]: nvm we are on python v3.10

**Review by kuhar — APPROVED** [review]

**Pull request merged**